File: 23061/23061.cpp
#include<iostream>
#include<algorithm>
#include<vector>
#define MAX 1000001
using namespace std;
int N, M;
int ans[101][MAX];
double maxValue;
int num=1;
vector<pair<int, int>> list;
vector<int> bag;
int main()
{
    ios::sync_with_stdio(false); cin.tie(0);
    cout.tie(0);
    cin >> N >> M;
    for (int i = 0; i < N;i++){
        int x, y; cin>>x>>y;
        list.push_back({x, y});
    }
    for (int i = 0; i < M;i++){
        int x; cin>>x;
        bag.push_back(x);
    }
    for (int i = 0; i < N; i++)
    {
        int w = list[i].first;
        int v = list[i].second;
        for (int j = 0; j < MAX; j++)
        {
            if (w <= j)
            {
                ans[i + 1][j] = max(ans[i][j], ans[i][j - w] + v);
            }
            else
            {
                ans[i + 1][j] = ans[i][j];
            }
        }
    }
    for (int i = 1; i < MAX;i++){
        ans[N][i] = max(ans[N][i], ans[N][i - 1]);
    }
    for (int i = 0; i < M;i++){
        int w = bag[i];
        double curValue = ans[N][w]*1.0 / w;
        if(curValue>maxValue){
            maxValue=curValue;
            num = i + 1;
        }
    }
    cout << num << '\n';
}

File: 23061/23061.js
var fs = require("fs");
var input = fs.readFileSync("input.txt").toString().split("\n");
const [N, M] = input[0].split(" ").map((v) => +v);
let list = [];
let bag = [];
let maxWeight = 1000000;
let ans = [];
let num = 0;
let maxValue = 0;
for (let i = 1; i <= N; i++) {
    const [w, v] = input[i].split(" ").map((v) => +v);
    list.push([w, v]);
}
for (let i = N + 1; i < N + M + 1; i++) {
    bag.push(+input[i]);
}
for (let i = 0; i <= N; i++) {
    ans.push([]);
    for (let j = 0; j <= maxWeight; j++) {
        ans[i][j] = 0;
    }
}
for (let i = 0; i < N; i++) {
    const [w, v] = list[i];
    for (let j = 0; j <= maxWeight; j++) {
        if (w <= j) {
            ans[i + 1][j] = Math.max(ans[i][j], ans[i][j - w] + v);
        } else {
            ans[i + 1][j] = ans[i][j];
        }
    }
}
for (let i = 1; i < maxWeight; i++) {
    ans[N][i] = Math.max(ans[N][i], ans[N][i - 1]);
}
for (let i = 0; i < M; i++) {
    const w = bag[i];
    const curValue = ans[N][w] / w;
    if (curValue > maxValue) {
        maxValue = curValue;
        num = i + 1;
    }
}
console.log(num);
